File: Backend-ProyectoWeb/src/main/java/org/web/project/service/AsociationService.java
package org.web.project.service;

import java.util.List;

import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;
import org.web.project.model.Asociation;
import org.web.project.model.Bus;
import org.web.project.model.Driver;
import org.web.project.model.Route;
import org.web.project.model.Schedule;
import org.web.project.repository.AsociationRepository;

@Service
public class AsociationService {
    
    @Autowired
    private AsociationRepository asociationRepository;


    public List<Asociation> getAllAsociations() {
        return asociationRepository.findAll();
    }

    public Asociation findAsociation(Long id) {
        return asociationRepository.findById(id).orElseThrow();
    }

    public Asociation saveAssignment(Asociation asociation) {
        return asociationRepository.save(asociation);
    }

    public List<Schedule> findSchedulesByRoute(Route route){
       return asociationRepository.findSchedulesByRoute(route);
    }

    public List<Route> findRouteByBus(Bus bus){
        return asociationRepository.findRoutesByBus(bus);
     }

     public List<Bus> findBusByDriver(Driver driver){
        return asociationRepository.findBusesByDriver(driver);
     }

     public  List<Bus> getAssignedBusesByRoute(Route route){
        return asociationRepository.getAssignedBusesByRoute(route);
     }

     public List<Driver> getAssignedDriversByBus(Bus bus){
        return asociationRepository.getAssignedDriversByBus(bus);
     }

     public List<Schedule> findSchedulesByBusAndRoute(Bus bus, Route route){
      return asociationRepository.findSchedulesByBusAndRoute(bus, route);
     }

      public List<Schedule> findSchedulesByDriver(Driver driver,Bus bus, Route route){
         return asociationRepository.findSchedulesByDriver(driver,bus,route);
     }
   

}

File: Frontend-ProyectoWeb/src/app/services/asociation.service.ts
import { Injectable } from '@angular/core';
import { Assignment } from '../model/assignment';
import { HttpClient, HttpHeaders } from '@angular/common/http';
import { Observable } from 'rxjs';
import { Schedule } from '../model/schedule';
import { Route } from '../model/route';
import { Bus } from '../model/bus';
import { Driver } from '../model/driver';

@Injectable({
  providedIn: 'root'
})
export class AsociationService {

  private httpOptions = {
    headers: new HttpHeaders({
      "Content-Type": "application/json"
    })
  };

  constructor(private http: HttpClient) { }

  findAll(): Observable<Assignment[]> {
    return this.http.get<Assignment[]>(`http://localhost:8080/api/assignment/list`)
  }

  update(id:number, asociation: Assignment): Observable<Assignment> {
    return this.http.put<Assignment>(`http://localhost:8080/api/assignment/update/${id}`, asociation, this.httpOptions);
  }

  save(asociation: Assignment): Observable<Assignment>{
   return this.http.post<Assignment>(`http://localhost:8080/api/assignment`, asociation, this.httpOptions);
  }

  findSchedulesByRoute(idRoute:number): Observable<Schedule[]> {
    return this.http.get<Schedule[]>(`http://localhost:8080/api/assignment/${idRoute}/schedules`)
  }

  findRoutesByBus(idBus:number): Observable<Route[]> {
    return this.http.get<Route[]>(`http://localhost:8080/api/assignment/${idBus}/routes`)
  }

  findBusesByDriver(idDriver:number): Observable<Bus[]> {
    return this.http.get<Bus[]>(`http://localhost:8080/api/assignment/${idDriver}/buses`)
  }

  getAssignedBusesByRoute(idRoute:number): Observable<Bus[]> {
    return this.http.get<Bus[]>(`http://localhost:8080/api/assignment/${idRoute}/assignedBuses`)
  }

  getAssignedDriversByBus(idBus:number): Observable<Driver[]> {
    return this.http.get<Driver[]>(`http://localhost:8080/api/assignment/${idBus}/assignedDrivers`)
  }

  findSchedulesByBusAndRoute(idBus: number,idRoute:number): Observable<Schedule[]> {
    return this.http.get<Schedule[]>(`http://localhost:8080/api/assignment/bus/${idBus}/route/${idRoute}`)
  }

  findSchedulesByDriver(idDriver:number,idBus:number,idRoute:number): Observable<Schedule[]> {
    return this.http.get<Schedule[]>(`http://localhost:8080/api/assignment/driver/${idDriver}/bus/${idBus}/route/${idRoute}`)
  }
}
